mgard: don't restrict protobuf version more than necessary

aumuell opened the pull request:
successfully built:
mgard@2022-11-18 ^protobuf@3.{4,21,25}
mgard@2023-01-10 ^protobuf@3.{4,25}
mgard@2023-03-31 ^protobuf@3.{4,25}

compile failures:
mgard@2022-11-18 ^protobuf@3.3
mgard@2023-01-10 ^protobuf@3.3
mgard@2023-03-31 ^protobuf@3.3

aumuell committed:
mgard: don't restrict protobuf version more than necessary

successfully built:
mgard@2022-11-18 ^protobuf@3.{4,21,25}
mgard@2023-01-10 ^protobuf@3.{4,25}
mgard@2023-03-31 ^protobuf@3.{4,25}

compile failures:
mgard@2022-11-18 ^protobuf@3.3
mgard@2023-01-10 ^protobuf@3.3
mgard@2023-03-31 ^protobuf@3.3

robertu94 commented:
Can you please share the compiler errors?

aumuell commented:
[linux-pb3.3-mgard-2022-11.txt](https://github.com/spack/spack/files/14594087/linux-pb3.3-mgard-2022-11.txt)
[linux-pb3.3-mgard-2023-01.txt](https://github.com/spack/spack/files/14594088/linux-pb3.3-mgard-2023-01.txt)
[linux-pb3.3-mgard-2023-03.txt](https://github.com/spack/spack/files/14594091/linux-pb3.3-mgard-2023-03.txt)

[macos-pb3.3-mgard-2022-11.txt](https://github.com/spack/spack/files/14594145/macos-pb3.3-mgard-2022-11.txt)
[macos-pb3.3-mgard-2023-01.txt](https://github.com/spack/spack/files/14594093/macos-pb3.3-mgard-2023-01.txt)
[macos-pb3.3-mgard-2023-03.txt](https://github.com/spack/spack/files/14594094/macos-pb3.3-mgard-2023-03.txt)

While I did not look into it, the errors on macOS with protobuf@3.3 and mgard@2022-11-18 are most likely caused by size_t and unsigned long or so being the same type - and they happen with any version of protobuf. But the other errors seem to come from incompatibilities with protobuf.

robertu94 commented:
I confirm this builds locally, but I see build failures in CI for powerpc and arm (which I don't have locally), but these failures could be transient due to out of memory errors since MGARD takes nearly 4GB/core to compile.  @alalazo can we re-run CI to confirm?

alalazo commented:
@spackbot run pipeline

robertu94 commented:
Ok, the ARM build isn't a transient.  Apparently, there is a bug in how ARM neon (ARM's version of vector instructions) gets used (I'm guessing by the protobuf generated source).  The powerpc build got a lot further, but still segfaulted when compiling the CUDA files that take an insane amount of RAM.

To be on the safe side, @aumuell we need to restrict this change to non-powerpc and non-arm based builds.  I think we can accomplish this by adding back a dependency on `"protobuf@3.21:" when="% ppc64le` and `"protobuf@3.21:" when="% arm`

aumuell committed:
mgard: add conflicts to address CI errors

aumuell committed:
mgard: conflict between cuda and abseil@20240116.1

compiling mgard+cuda with gcc@12.3.0 and nvcc from cuda@12.3.0 against
protobuf pulling in abseil-cpp@20240116.1 results in the errors reported
here: https://github.com/abseil/abseil-cpp/issues/1629

aumuell commented:
I tried to address the CI issues by adding two `conflicts`.

While testing compilation with `+cuda` locally, I ran into incompatibilities between abseil@20240116.1 and nvcc from cuda@12.3.0, as also reported here: https://github.com/abseil/abseil-cpp/issues/1629 . So I added a conflict for that as well.
